website/blog/2023-01-24-aggregating-test-failures.md

```diff
@@ -0,0 +1,134 @@
+
+Recently, we overhauled our testing framework. We cut the number of tests down to 200, creating a more mature framework that includes metadata and emphasizes actionability. Our system for managing data quality is a three step process, described below:
+
+1. Leveraging the contextual knowledge of stakeholders, writing specific, high quality data tests, perpetuating test failure results into aliased models for easy access. 
+1. Aggregating test failure results using Jinja macros and pre-configured metadata to pull together high level summary tables. 
+1. Building views on top of the base table to split tests by owner or severity, and creating visualizations using our tool of choice.
+
+_It should be noted that this framework is for dbt v1.0+ on BigQuery. Small adaptations are likely necessary to make this framework run on a different setup._
+
+## Specific, high quality data tests
+
+When we talk about high quality data tests, we aren’t just referencing high quality code, but rather the informational quality of our testing framework and their corresponding error messages. Originally, we theorized that any test that cannot be acted upon is a test that should not be implemented. Later, we realized there is a time and place for tests that should receive attention at a critical mass of failures. All we needed was a higher specificity system: tests should have an explicit severity ranking associated with them, equipped to filter out the noise of common, but low concern, failures. Each test should also mesh into established [RACI](https://project-management.com/understanding-responsibility-assignment-matrix-raci-matrix/) guidelines that state which groups tackle what failures, and what constitutes a critical mass.
+
+To ensure that tests are always acted upon, we implement tests differently depending on the user groups that must act when a test fails. This led us to have two main classes of tests — Data Integrity Tests (called [Generic Tests](https://docs.getdbt.com/docs/build/tests) in dbt docs) and Context Driven Tests (called [Singular Tests](https://docs.getdbt.com/docs/build/tests#singular-tests) in dbt docs), with varying levels of severity across both test classes.
+
+Data Integrity tests (Generic Tests)  are simple — they’re tests akin to a uniqueness check or not null constraint. These tests are usually actionable by the data platform team rather than subject matter experts. We define Data Integrity tests in our YAML files, similar to how they are [outlined by dbt’s documentation on generic tests](https://docs.getdbt.com/docs/build/tests). They look something like this —
+
+```yaml
+version: 2
+models:
+  - name: patient
+    columns:
+      - name: id
+        description: Unique ID associated with the record
+        tests:
+          - unique:
+              alias: patient_ _id_ _unique
```

Comment:
Could we remove the space between the underscores here (e.g. `_ _` to `__` and on line 48? that was a copy-paste error on my side - you can see double underscores in code previews but not plain text.

Reply:
Just adjusted this, please lmk if this works for you!